File: benchmarks/quicksort/c/quicksort.c
#include <stdio.h>
#include <stdlib.h>
#include <inttypes.h>

/*
 * In-place Lomuto-partition quicksort over a heap-allocated i32 array.
 * Heap allocation (not a stack array) is deliberate and applies uniformly
 * across all 4 languages in this benchmark: n can reach into the millions,
 * and a stack array that size would overflow the default 8MB thread stack
 * in any of C/Nim/Zig/yap alike, not just here.
 */
static void swap_elem(int32_t *arr, int32_t i, int32_t j) {
    int32_t tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
}

static int32_t partition(int32_t *arr, int32_t lo, int32_t hi) {
    int32_t pivot = arr[hi];
    int32_t i = lo - 1;
    for (int32_t j = lo; j < hi; j++) {
        if (arr[j] <= pivot) {
            i++;
            swap_elem(arr, i, j);
        }
    }
    swap_elem(arr, i + 1, hi);
    return i + 1;
}

static void quicksort(int32_t *arr, int32_t lo, int32_t hi) {
    if (lo < hi) {
        int32_t p = partition(arr, lo, hi);
        quicksort(arr, lo, p - 1);
        quicksort(arr, p + 1, hi);
    }
}

/* Deterministic pseudo-shuffled fill, identical formula across all 4 languages. */
static void fill(int32_t *arr, int32_t n) {
    for (int32_t i = 0; i < n; i++) {
        arr[i] = (int32_t)(((uint64_t)(i + 1) * 2654435761ULL) % 1000003ULL);
    }
}

static uint64_t checksum(const int32_t *arr, int32_t n) {
    uint64_t sum = 0;
    for (int32_t i = 0; i < n; i++) {
        sum += (uint64_t)arr[i] * (uint64_t)(i + 1);
    }
    return sum;
}

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <n>\n", argv[0]);
        return 1;
    }
    int32_t n = (int32_t)strtol(argv[1], NULL, 10);

    if (n <= 0) {
        printf("%" PRIu64 "\n", (uint64_t)0);
        return 0;
    }

    int32_t *arr = malloc((size_t)n * sizeof(int32_t));
    if (!arr) {
        fprintf(stderr, "malloc failed\n");
        return 1;
    }

    fill(arr, n);
    quicksort(arr, 0, n - 1);
    uint64_t result = checksum(arr, n);
    free(arr);

    printf("%" PRIu64 "\n", result);
    return 0;
}

File: benchmarks/quicksort/zig/quicksort.zig
const std = @import("std");

fn swapElem(arr: []i32, i: i32, j: i32) void {
    const tmp = arr[@intCast(i)];
    arr[@intCast(i)] = arr[@intCast(j)];
    arr[@intCast(j)] = tmp;
}

fn partition(arr: []i32, lo: i32, hi: i32) i32 {
    const pivot = arr[@intCast(hi)];
    var i = lo - 1;
    var j = lo;
    while (j < hi) : (j += 1) {
        if (arr[@intCast(j)] <= pivot) {
            i += 1;
            swapElem(arr, i, j);
        }
    }
    swapElem(arr, i + 1, hi);
    return i + 1;
}

fn quicksort(arr: []i32, lo: i32, hi: i32) void {
    if (lo < hi) {
        const p = partition(arr, lo, hi);
        quicksort(arr, lo, p - 1);
        quicksort(arr, p + 1, hi);
    }
}

fn fill(arr: []i32, n: i32) void {
    var i: i32 = 0;
    while (i < n) : (i += 1) {
        const idx: u64 = @intCast(i + 1);
        arr[@intCast(i)] = @intCast((idx *% 2654435761) % 1000003);
    }
}

fn checksum(arr: []i32, n: i32) u64 {
    var sum: u64 = 0;
    var i: i32 = 0;
    while (i < n) : (i += 1) {
        const v: u64 = @intCast(arr[@intCast(i)]);
        const idx: u64 = @intCast(i + 1);
        sum +%= v *% idx;
    }
    return sum;
}

pub fn main(init: std.process.Init) !void {
    var it = init.minimal.args.iterate();
    _ = it.skip(); // program name

    const n_str = it.next() orelse {
        std.debug.print("Usage: quicksort <n>\n", .{});
        std.process.exit(1);
    };

    const n = try std.fmt.parseInt(i32, n_str, 10);

    var buf: [32]u8 = undefined;
    var stdout = std.Io.File.stdout().writer(init.io, &buf);

    if (n <= 0) {
        try stdout.interface.print("{d}\n", .{@as(u64, 0)});
        try stdout.interface.flush();
        return;
    }

    const allocator = std.heap.page_allocator;
    const arr = try allocator.alloc(i32, @intCast(n));
    defer allocator.free(arr);

    fill(arr, n);
    quicksort(arr, 0, n - 1);
    const result = checksum(arr, n);

    try stdout.interface.print("{d}\n", .{result});
    try stdout.interface.flush();
}
